### PR TITLE
~Hotfix~ (coldfix) file size exploit && file resume storage info.

### PR DESCRIPTION
**Pull request by GrayHatter**
There was a risk of a friend being able to start a file to us, have us accept the file. Then if the friend was to go disconnect, and wait for our client to restart. That friend would be able to send a file of an arbitrary length, and we would accept it without question.

Additionally, there was a max limit of 32 resumable transfers, for all friends combined. Now the only limit is the of the max concurrent transfers PER friend. So that one friend won't interfere with a different friends transfers.

Finally; previously, there was a risk of data leakage as outgoing transfers only searched by friend number. If any lower number friend was removed from the list, or the original friend was removed. uTox would try to resume a file to which ever friend took that friends spot. Now if a friend number is changed, uTox will only resume a file if the pubkey matches for that friend, regardless of the number. Small caveat, if you remove that friend with a pending broken transfer, the information to resume that transfer will remain on disk.

@irungentoo I've tested this and it's ready to merge.

### CODE REVIEW

File: file_transfers.c [hunk 1]
```diff
@@ -509,94 +482,123 @@ static void incoming_file_avatar(Tox *tox, uint32_t friend_number, uint32_t file
...
+        }
+    } else {
+        /* File number wasn't the same, maybe it's still somewhere? */
+        int good = 0; /* Used to skip searching for files */
+        FILE_TRANSFER *file_search;
+        for(int i = 0 ; i < MAX_FILE_TRANSFERS ; i++){
+            file_search = get_file_transfer(friend_number, i);
+            if(memcmp(file_search->file_id, file_id, TOX_FILE_ID_LENGTH) == 0){
+                memcpy(file_handle, file_search, sizeof(*file_handle));
+                file_handle->file_number = file_number;
```
Comment (irungentoo) [inline file_transfers.c]: good = 1;

?

Comment (GrayHatter) [inline file_transfers.c]: Dirty hack to find out if uTox was able to find that file in memory with a
different file number

**Comment by irungentoo** [comment]
Your code is a bit of a mess.

Here's what you should be doing:

For incoming files:
1. Every time the user accepts a file the file info is stored in a new file: [file_id].file_in
2. If the file is canceled/finished you remove that file.
3. For every new incoming files you check if that file ([file_id].file_in) exists, if it does you resume, if it doesn't, it's a new file transfer.
4. When the friend goes offline you make uTox clear all the data it has on the incoming files tied to that friend in memory (you zero out all the incoming_transfer[MAX_NUM_FRIENDS][MAX_FILE_TRANSFERS] entry for the friend.).

That way you don't need to loop and the code will be smaller and simpler than it is now.

For outgoing files:
1. For every new outgoing file, (right after it is created) you save its info to a [friend public key][index].file_out kinda how you do now.
2. If the file is canceled/finished you remove that file.
3. When the friend goes offline you make uTox clear all the data it has on the outgoing files tied to that friend in memory (you zero out all the outgoing_transfer[MAX_NUM_FRIENDS][MAX_FILE_TRANSFERS] entry for the friend.)
4. When a friend goes online you try loading all the .file_out files for the friend and send the files corresponding to them.

**Comment by irungentoo** [comment]
I get a errors when I send 2 files at the same time from both sides then restart one of the clients while the files are transferring.

**Comment by GrayHatter** [comment]
replaced by #993

**Pull request closed without merging**